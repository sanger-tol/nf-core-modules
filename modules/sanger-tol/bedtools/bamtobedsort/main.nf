process BEDTOOLS_BAMTOBEDSORT {
    tag "$meta.id"
    label "process_high"

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://community-cr-prod.seqera.io/docker/registry/v2/blobs/sha256/cf/cf65e22485bc417775bef295e324b214d5a4ddfea5c3cbfedf8623bf8af55612/data' :
        'community.wave.seqera.io/library/bedtools_samtools_coreutils:43e34cbcf7cfff84' }"

    input:
    tuple val(meta), path(bam)

    output:
    tuple val(meta), path("*.bed"), emit: sorted_bed
    path "versions.yml"           , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix      = task.ext.prefix ?: "${meta.id}"
    def args1       = task.ext.args1  ?: ""
    def args2       = task.ext.args2  ?: ""
    def args3       = task.ext.args3  ?: ""
    def st_cores    = task.cpus > 4 ? 4 : "${task.cpus}"
    def buffer_mem  = (task.memory.toGiga() / 2).round()
    """
    samtools view \\
        -@${st_cores} \\
        ${args1} \\
        ${bam} | \\
    bamToBed ${args2} -i stdin | \\
    sort ${args3} \\
        --parallel=${task.cpus} \\
        -S ${buffer_mem}G \\
        -T . > \\
    ${prefix}.bed

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
        bedtools: \$(bedtools --version | sed -e "s/bedtools v//g")
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.bed

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
        bedtools: \$(bedtools --version | sed -e "s/bedtools v//g")
    END_VERSIONS
    """
}
