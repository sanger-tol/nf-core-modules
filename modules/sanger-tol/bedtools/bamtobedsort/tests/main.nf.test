
nextflow_process {

    name "Test Process BEDTOOLS_BAMTOBEDSORT"
    script "../main.nf"
    process "BEDTOOLS_BAMTOBEDSORT"

    tag "modules"
    tag "modules_sangertol"
    tag "bedtools"
    tag "bedtools/bamtobedsort"

    config './nextflow.config'

    test("Undibacterium unclassified - bam") {

        when {

            params {
                samtools_view_args = "-u -F0x400"
                sort_args          = "-k4"
            }

            process {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test' ], // meta map
                        file(params.modules_testdata_base_path + 'Undibacterium_unclassified/analysis/hic_mapping/baUndUnlc1_hic_phased_hap2_merged.bam', checkIfExists: true)
                    ]
                )
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("Undibacterium unclassified - bam - stub") {

        options "-stub"

        when {

            params {
                samtools_view_args = "-u -F0x400"
                sort_args          = "-k4"
            }

            process {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test' ], // meta map
                        file(params.modules_testdata_base_path + 'Undibacterium_unclassified/analysis/hic_mapping/baUndUnlc1_hic_phased_hap2_merged.bam', checkIfExists: true)
                    ]
                )
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
