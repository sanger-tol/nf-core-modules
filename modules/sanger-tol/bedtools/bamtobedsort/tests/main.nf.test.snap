{
    "Undibacterium unclassified - bam": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.bed:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,bebe088109df718929558eced700f1f4"
                ],
                "sorted_bed": [
                    [
                        {
                            "id": "test"
                        },
                        "test.bed:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,bebe088109df718929558eced700f1f4"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-08-22T15:32:16.253374"
    },
    "Undibacterium unclassified - bam - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.bed:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,3791cfd84a437fe3de62017960a5745f"
                ],
                "sorted_bed": [
                    [
                        {
                            "id": "test"
                        },
                        "test.bed:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,3791cfd84a437fe3de62017960a5745f"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-08-22T15:32:18.312679"
    }
}